# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

1. Выведем нужные для анализа данных столбцы 'children', 'debt', чтобы взглянуть и понять, что из себя представляют данные.

In [22]:
data[['children', 'debt']]

children  debt
0             1     0
1             1     0
2             0     0
3             3     0
4             0     0
...         ...   ...
21520         1     0
21521         0     0
21522         1     1
21523         3     1
21524         2     0

[21331 rows x 2 columns]

2. Выведем на экран уникальные значения с количеством детей, чтобы проверить, нет ли аномалий в наших данных.

In [23]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

3. Проверим, чтобы в обоих столбцах не было пустых ячеек. Это мы делаем, для того, чтобы анализ был максимально точным.

In [24]:
data['children'].isna().sum()

0

In [25]:
data['debt'].isna().sum()

0

4. Создадим функцию, которая будет разделять на категории, в зависимости от количестваа детей в семье. Если детей нет - категория "Нет детей". Если детей в семье 1-2 - категория "Есть дети". Если в семье 3 и более ребенка - категория "Многодетный".

In [26]:
def children_category(children):
    if 1 <= children <= 2:
        return 'Есть дети'
    if children >= 3:
        return 'Многодетный'
    if children == 0:
        return 'Нет детей'

5. Создадим отдельный столбец в таблице, в котором будет информация о категориях в зависимости от количества детей. Воспользуемся функцией apply() для того, чтобы она применила функцию 'children_category'.

In [27]:
data['children_category'] = data['children'].apply(children_category)


6. Для более глубокого анализа создадим сводную таблицу.

In [28]:
debt_children = pd.pivot_table(data,
              index = ['children_category'], # По какому признаку у нас будет осуществляться агрегация данных
              columns = ['debt'], # Определим явно, какие нам нужны для анализа столбцы.
              values = ['gender'], # Определяем значение 'gender' для дальнейшей замены 0 и 1 на no_debt и debt.
              aggfunc = 'count') # Агрегированная функция, которая считывает все значения в стобцах.
debt_children.columns = ['no_debt', 'debt'] # Создаем два столбца недолжник / должник, заменяя значения 'gender' 0 и 1.
debt_children

no_debt  debt
children_category               
Есть дети             6222   638
Многодетный            349    31
Нет детей            13028  1063

7. По выведенным данным можно уже сделать вывод, но для наглядности выведем еще процентное соотношение между должником и недолжником.

In [29]:
debt_children['interest'] = (debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt']))
debt_children['interest'] = debt_children['interest'].map('{:.1%}'.format) # Вывод в процентном формате.
debt_children.sort_values(by = 'interest', ascending = True) # Сортировка по столбцу 'interest' по убыванию.
debt_children

no_debt  debt interest
children_category                        
Есть дети             6222   638     9.3%
Многодетный            349    31     8.2%
Нет детей            13028  1063     7.5%

**Вывод:**

Из приведенной выше таблицы мы можем понять, что семьянины, у которых 1-2 ребенка, являются самыми рисковыми для банка клиентами. По нашим подсчетам 9.3% от общего числа из категории 'Есть дети' - должники. На втором месте идет категория "Многодетный" (8.2%), хоть детей больше, чем в предыдущей группе, но процент должников меньше, скорей всего, это связано с тем, что государство помогает многодетным семьям. И те, люди, у которых нет детей (7.5%), являются менее рисковыми для банка. Из этого вывода, мы можем понять, что показатель возврата кредита в срок зависит от количества детей (наличие детей повышает вероятность задолженности), но стоит учесть, что  государство даёт доп.субсидии многодетным семьям.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

1. Выведем нужные для анализа данных столбцы 'family_status', 'debt', чтобы взглянуть и понять, что из себя представляют данные.

In [30]:
data[['family_status', 'debt']]

family_status  debt
0       женат / замужем     0
1       женат / замужем     0
2       женат / замужем     0
3       женат / замужем     0
4      гражданский брак     0
...                 ...   ...
21520  гражданский брак     0
21521   женат / замужем     0
21522  гражданский брак     1
21523   женат / замужем     1
21524   женат / замужем     0

[21331 rows x 2 columns]

2. Выведем уникальные значения столбца 'family_status', чтобы проверить нет ли дубликатов значений.

In [31]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

3. Проверим нет ли пропусков в столбцах, которые нам нужны для анализа.

In [32]:
data['family_status'].isna().sum()

0

In [33]:
data['debt'].isna().sum()

0

4. Создадим сводную таблицу, в которой главным столбцом будет 'family_status' и два столбца 'no_debt' и 'debt', показывающие должник или недолжник наш клиент.

In [34]:
family_st_debt = pd.pivot_table(data,
                               index = ['family_status'], # По какому признаку у нас будет осуществляться агрегация данных
                               columns = ['debt'], # Определим явно, какие нам нужны для анализа столбцы.
                               values = ['gender'], # Определяем значение 'gender' для дальнейшей замены 0 и 1 на no_debt и debt.
                               aggfunc = 'count') # Агрегированная функция, которая считывает все значения в стобцах.
family_st_debt.columns = ['no_debt', 'debt'] # Создаем два столбца недолжник / должник, заменяя значения 'gender' 0 и 1.и
family_st_debt

no_debt  debt
family_status                       
Не женат / не замужем     2523   273
в разводе                 1105    84
вдовец / вдова             888    63
гражданский брак          3749   385
женат / замужем          11334   927

5. Создадим новый столбец 'interest', в котором будет изображен процент соотношения дожников на общее количество в определенном семейном статусе.

In [35]:
family_st_debt['interest'] = (family_st_debt['debt'] / (family_st_debt['debt'] + family_st_debt['no_debt'])) 
family_st_debt['interest'] = family_st_debt['interest'].map('{:.1%}'.format) # Вывод в процентном формате.
family_st_debt.sort_values(by = 'interest', ascending = False) # Сортировка по столбцу 'interest' по убыванию.


no_debt  debt interest
family_status                                
Не женат / не замужем     2523   273     9.8%
гражданский брак          3749   385     9.3%
женат / замужем          11334   927     7.6%
в разводе                 1105    84     7.1%
вдовец / вдова             888    63     6.6%

**Вывод:**

Из приведенной выше таблице, мы видим, что самый рисковый клиент для выдачи кредитов является неженатые люди (±9.8%) - они имеют самый высокий процент по задолженностям среди других, это можно объяснить тем, что неженатые люди, чаще всего не имеют чувства ответственности, поэтому делают легкомысленные покупки, не имея возможности отдать долг. На втором месте по проценту задолженности - люди в гражданском браке (±9.3%). На третьем месте расположились - женатые (7.6%). Четвертое - люди в разводе (7.1%). Пятое - вдовец / вдова (6.6%).



#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

1. Выведем нужные для анализа данных столбцы 'total_income', 'debt', чтобы взглянуть и понять, что из себя представляют данные.

In [36]:
data[['total_income', 'debt']]

total_income  debt
0            253875     0
1            112080     0
2            145885     0
3            267628     0
4            158616     0
...             ...   ...
21520        224791     0
21521        155999     0
21522         89672     1
21523        244093     1
21524         82047     0

[21331 rows x 2 columns]

2. Проверим, чтобы в обоих столбцах не было пропущенных ячеек. Это мы делаем, для того, чтобы анализ был максимально точным.

In [37]:
data['total_income'].isna().sum()

0

In [38]:
data['debt'].isna().sum()

0

3. Создаем функцию 'income_category' для того, чтобы разделить доход на несколько категорий 'A', 'B', 'C', 'D', 'E'. Категория 'A' - до 30000, 'B' - от 30001 до 50000, 'C' - от 50001 до 200000, 'D' - от 200001 до 1000000, 'E' - больше 1000000.


In [39]:
def income_category(total_income):
    
    if 0 <= total_income <= 30000:
        return 'E'
    elif 30001 <= total_income <= 50000:
        return 'D'
    elif 50001 <= total_income <= 200000:
        return 'C'
    elif 200001 <= total_income <= 1000000:
        return 'B'
    elif total_income >= 1000001:
        return 'A'


4. Создадим отдельный столбец в таблице, в котором будет информация о категориях в зависимости от уровня дохода. Воспользуемся функцией apply() для того, чтобы она применила функцию 'income_category'.

In [40]:
data['income_category'] = data['total_income'].apply(income_category)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category children_category income_category  
0      операции с недвижимостью         Есть дети               B  
1        операции с автомобилем         Есть дети               C  
2      операции с недвижимостью         Нет детей               C  
3         получение образования       Многодетный               B  
4            проведение свадьбы         Нет детей               C  
...                         ...               ...             ...  
21520  операции с недвижимостью         Есть дети               B  
21521    операции с автомобилем         Нет детей               C  
21522  операции с недвижимостью         Есть дети               C  
21523    операции с автомобилем       Многодетный               B  
21524    операции с автомобилем         Есть дети               C  

[21331 rows x 16 columns]

5. Создадим сводную таблицу, в которой главным столбцом будет 'income_category' и два столбца 'no_debt' и 'debt', показывающие должник или недолжник наш клиент.

In [41]:
income_debt = pd.pivot_table(data,
                            index = ['income_category'], # По столбцу 'income_category' у нас будет осуществляться агрегация данных
                            columns = ['debt'], # Определим явно, какие нам нужны для анализа столбцы.
                            values = ['gender'], # Определяем значение 'gender' для дальнейшей замены 0 и 1 на no_debt и debt.
                            aggfunc = 'count') # Агрегатная функция, подсчет всех значений.
income_debt.columns = ['no_debt', 'debt'] # Создание двух столбцов 'no_debt' и 'debt' (вместо 0 и 1)
                            

income_debt

no_debt  debt
income_category               
A                     23     2
B                   4660   354
C                  14568  1353
D                    328    21
E                     20     2

6. Создадим новый столбец 'interest', в котором будет изображен процент соотношения дожников на общее количество в определенной категории по уровню дохода.

In [42]:
income_debt['interest'] = (income_debt['debt'] / (income_debt['debt'] + income_debt['no_debt']))
income_debt['interest'] = income_debt['interest'].map('{:.1%}'.format) # Вывод в процентном формате.
income_debt.sort_values(by = 'interest', ascending = False) # Сортировка по убыванию по столбцу 'interest'

no_debt  debt interest
income_category                        
E                     20     2     9.1%
C                  14568  1353     8.5%
A                     23     2     8.0%
B                   4660   354     7.1%
D                    328    21     6.0%

**Вывод:** 

Из приведенной выше таблицы, мы можем увидеть, что у строк с категориями 'E', 'A', 'D' мало данных по отношению с другими категориями. Необходимо проверить правильность переноса данных в таблицу из базы данных, задать вопрос разработчикам, действительно ли столь мало людей берут кредит из данных категорий. Поэтому, пока что, мы опустим показатели этих категорий и возьмем для анализа данные из категорий 'C' и 'B'. Из таблицы видно, что категория ('C') людей с заработком от 50001 - 200000 является наиболее подверженой риску быть должниками. На втором месте идут люди получающие от 200001 до 1000000 (категория 'B').


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

1. Для того, чтобы понимать, с какими данными нам придется работать, выведем оба столбца 'purpose' - цель взятия кредита, 'dept' - имел ли наш клиент задолженность по возврату кредитов до этого.

In [43]:
data[['purpose', 'debt']]

purpose  debt
0                     покупка жилья     0
1           приобретение автомобиля     0
2                     покупка жилья     0
3        дополнительное образование     0
4                   сыграть свадьбу     0
...                             ...   ...
21520             операции с жильем     0
21521          сделка с автомобилем     0
21522                  недвижимость     1
21523  на покупку своего автомобиля     1
21524         на покупку автомобиля     0

[21331 rows x 2 columns]

2. Проверим целостность данных в обоих столбцах (проверим пропуски), для того чтобы анализ был наиболее точный.

In [44]:
data['purpose'].isna().sum()

0

In [45]:
data['debt'].isna().sum()

0

3. Выведем все уникальные значения в столбце 'purpose', чтобы увидеть основные цели для взятия кредита.

In [46]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

3. Из выше приведенного списка, можно понять, что цели похожи, поэтому сформируем все значения в столбце 'purpose' по категориям взятия кредита с помощью функции categorize_purpose. Категории: 'Покупка/строительство недвижимости', 'Покупка автомобиля', 'Организация свадьбы', 'Получение образования'.

In [47]:
def categorize_purpose(purpose):
    try:
        if 'авто' in purpose:
            return 'Покупка автомобиля'
        if 'свадьб' in purpose:
            return 'Организация свадьбы'
        if 'образ' in purpose:
            return 'Получение образования'
        if 'жиль' in purpose or 'недв' in purpose:
            return 'Покупка/строительство недвижимости'
    except:
        return 'Нет категории'

4. Создадим отдельный столбец, в который будем выводить категории взятия кредита. Воспользуемся функцией apply() для того, чтобы она применила функцию 'children_category'.

In [48]:
data['category_purpose'] = data['purpose'].apply(categorize_purpose)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category children_category income_category  \
0      операции с недвижимостью         Есть дети               B   
1        операции с автомобилем         Есть дети               C   
2      операции с недвижимостью         Нет детей               C   
3         получение образования       Многодетный               B   
4            проведение свадьбы         Нет детей               C   
...                         ...               ...             ...   
21520  операции с недвижимостью         Есть дети               B   
21521    операции с автомобилем         Нет детей               C   
21522  операции с недвижимостью         Есть дети               C   
21523    операции с автомобилем       Многодетный               B   
21524    операции с автомобилем         Есть дети               C   

                         category_purpose  
0      Покупка/строительство недвижимости  
1                      Покупка автомобиля  
2      Покупка/строительство недвижимости  
3                   Получение образования  
4                     Организация свадьбы  
...                                   ...  
21520  Покупка/строительство недвижимости  
21521                  Покупка автомобиля  
21522  Покупка/строительство недвижимости  
21523                  Покупка автомобиля  
21524                  Покупка автомобиля  

[21331 rows x 17 columns]

5. Для того, чтобы проверить гипотезу воспользуемся функцией pivot_table(), для создания сводной таблицы, которая покажет нам данные.

In [49]:
purpose_debt = pd.pivot_table(data,
                             index = 'category_purpose', # По столбцу 'category_purpose' у нас будет осуществляться агрегация данных
                             columns = 'debt',  # Определим явно, какие нам нужны для анализа столбцы.
                             values = 'gender', # Определяем значение 'gender' для дальнейшей замены 0 и 1 на no_debt и debt.
                             aggfunc = 'count') # Агрегатная функция, подсчет всех значений.
purpose_debt.columns = ['no_debt', 'debt'] # Создание двух столбцов 'no_debt' и 'debt' (вместо 0 и 1)
purpose_debt                       

no_debt  debt
category_purpose                                 
Организация свадьбы                    2130   183
Покупка автомобиля                     3879   400
Покупка/строительство недвижимости     9971   780
Получение образования                  3619   369

6. Для того, чтобы приведенные данные были понятны, приведем процентное соотношение должников по отношению к общему количеству клиентов в каждой категории.

In [50]:
purpose_debt['interest'] = purpose_debt ['debt'] / (purpose_debt['debt'] + purpose_debt['no_debt'])
purpose_debt['interest'] = purpose_debt['interest'].map('{:.2%}'.format) # Вывод в процентном формате.
purpose_debt.sort_values(by = 'interest', ascending = False) # Сортировка столбца 'interest' по убыванию.


no_debt  debt interest
category_purpose                                          
Покупка автомобиля                     3879   400    9.35%
Получение образования                  3619   369    9.25%
Организация свадьбы                    2130   183    7.91%
Покупка/строительство недвижимости     9971   780    7.26%

**Вывод:** 

Из приведенной выше сводной таблицы можно понять, что самая рисковая категория цели на кредит - покупка автомобиля (9.35%) и получение платного образования (9.25%). Перед согласованием кредита, стоит более тщательно проверить людей, которые берут кредит по этим целям, смогут ли они выплатить этот кредит? Также, организация свадьбы (7.91%) и покупка/строительство (7.26%) наиболее надежные категории.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
- Человеческий фактор: ошибки при сборе даннных или ошибки при вводе данных в систему могут привести к пропускам, также к человеческому фактору можно добавить, что клиенты/пользователи могли не предоставить данные (необязательные поля, например, email).
- Технический фактор: был составлен неправильный запрос / объединение таблиц, а также пропуски могли возникнуть, если данные были сохранены в неправильном формате или были повреждены при передаче или сохранении.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Заполнить пропуски медианным значением - лучшее решение для количественных переменных, потому что в столбце могут встречаться значения, которые сильно отличаются от большего числа значений, и если мы будем заполнять пустые значения средним арифметическим - это будет некорректно, так как показатели будут сильно искажены (из-за аномалий в данных). А если мы будем заполнять пропуски медианным значением, то это позволит нам избежать искажений и проанализровать тщательно данные.

### Шаг 4: общий вывод.

В данном модуле мы изучили основные моменты, связанные с предобработкой данных: работа с пропусками, определение аномальных значений, преобразование типов данных, основные методы поиска дубликатов, работа с несовершенными реальными наборами данных. По окончании модуля мы сделали проект, в котором проанализировали влияние различных показателей друг на друга, и можем сделать вывод по ним: 
1) Наличие детей повышает вероятность задолженности.

2) Категория людей 'Не женат / не замужем' - являются рискованным клиентом для банка. 

3) Категория людей 'C' с заработком 50001 - 200000 - является наиболее рисковой для выдачи кредита.

4) Граждане берущие кредит для покупки автомобиля или платного обучения - наиболее подвержены риску стать должниками.

Из выше перечисленного вывода, мы можем составить портрет идеального и наиболее рискованного клиента для банка-заказчика.
- **Идеальный клиент для банка**: женатый/замужний человек, без детей, с заработком от 200001 до 1000000, берущий кредит на покупку/строительство недвижимости.
- **Наиболее рискованный клиент для банка:** неженатый/незамужний человек, 1-2 ребенка, с заработком от 50001 до 200000, берущий кредит на покупку автомобиля / оплату образования.